src/libstd/io/mod.rs

```diff
@@ -942,6 +942,55 @@ impl<'a> IoSliceMut<'a> {
     pub fn new(buf: &'a mut [u8]) -> IoSliceMut<'a> {
         IoSliceMut(sys::io::IoSliceMut::new(buf))
     }
+
+    /// Advance the internal cursor of the slice.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(io_slice_advance)]
+    ///
+    /// use std::io::IoSliceMut;
+    /// use std::ops::Deref;
+    ///
+    /// let mut buf1 = [1; 8];
+    /// let mut buf2 = [2; 16];
+    /// let mut buf3 = [3; 8];
+    /// let mut bufs = &mut [
+    ///     IoSliceMut::new(&mut buf1),
+    ///     IoSliceMut::new(&mut buf2),
+    ///     IoSliceMut::new(&mut buf3),
+    /// ][..];
+    ///
+    /// // Mark 10 bytes as read.
+    /// IoSliceMut::advance(&mut bufs, 10);
+    /// assert_eq!(bufs[0].deref(), [2; 14].as_ref());
+    /// assert_eq!(bufs[1].deref(), [3; 8].as_ref());
+    /// ```
+    #[unstable(feature = "io_slice_advance", issue = "62726")]
+    #[inline]
+    pub fn advance(bufs: &mut &mut [IoSliceMut<'a>], n: usize) {
```

Comment:
One other option for an API signature is:

```rust
pub fn advance(bufs: &'b mut [IoSliceMut<'a>], n: usize) -> &'b mut [IoSliceMut<'a>]
```

but that's perhaps a bit wonky since it still mutates the contents themselves. It's also somewhat wonky though to take `&mut &mut`, so I'm not really sure if there's a slam dunk one way or another.

Reply:
Hm can you elaborate why it wouldn't work? I'm not sure I understand what the limitation would be

Reply:
For some reason I didn't think that API would work when storing `&mut [IoSliceMut]` in a struct, but it does.

I do think your suggested API is nicer, however I do have one reservation wouldn't it be surprising that the original `IoSliceMut` is sometimes modified? People might expect that effectively a sub slice of the original `bufs` is returned even though they're passing in `&mut`. If you don't think its a problem I'll change the API.

Reply:
While I see them both as a bit wonky, I think I prefer the API that returns a slice with a doc comment indicating that the elements may be mutated.

Reply:
I've updated the API.